### DISCUSSION TITLE
Missing Origin Header

### DISCUSSION THREAD

**Discussion by yancahila** [AzureAD/microsoft-authentication-library-for-python#392] [Q&A] [answered]
Upon looking the codes, I was wondering if there's any way to specify additional headers aside from the default ones? Advance thanks for the insights. :)

oauth2.py
    def _obtain_token(  # The verb "obtain" is influenced by OAUTH2 RFC 6749
            self, grant_type,
            params=None,  # a dict to be sent as query string to the endpoint
            data=None,  # All relevant data, which will go into the http body
            headers=None,  # a dict to be sent as request headers
            post=None,  # A callable to replace requests.post(), for testing.
                        # Such as: lambda url, **kwargs:
                        #   Mock(status_code=200, text='{}')
            **kwargs  # Relay all extra parameters to underlying requests
            ):  # Returns the json object came from the OAUTH2 response
        _data = {'client_id': self.client_id, 'grant_type': grant_type}

I can't find a way for the headers parameter to be supplied from the outside. 

Btw, I'm using the available python code for Flask.

**Comment by rayluo** [comment]
We do not expose additional headers as parameter in MSAL's API surface. Why would you want to do so? Can you describe your scenario first, before we start discussing the implementation, so that we avoid the [XY problem](https://meta.stackexchange.com/a/66378)?

BTW, what did you mean by "using the available python code for Flask"? Are you using any of our flask-based samples? If so, please also leave its link so that we know specifically which one.

UPDATE: Future readers can skip the lengthy conversation and just head to the final solution [here](https://github.com/AzureAD/microsoft-authentication-library-for-python/discussions/392#discussioncomment-1256665).

**Reply by yancahila** [reply]
The repo can be found [here](https://github.com/Azure-Samples/ms-identity-python-webapp).

So my problem is that the server returns AADSTS9002327: Tokens issued for the 'Single-Page Application' client-type may only be redeemed via cross-origin requests everytime I attempt to validate a user. So from the error description, it has something to do with CORS and basically the fix was I only have to add the 'Origins' field in the headers in getting the token. Now, I'm trying to add this field from the outside and can't seem to find a way.

**Reply by yancahila** [reply]
I tried running this [repo](https://github.com/Azure-Samples/ms-identity-python-flask-webapp-authentication) as well and it returned the same thing, Is there anything I should configure on my end? Additional detail, I'm using App Registrations in configuring my app and the org's AD. 

Here's the snap of the error,
![image](https://user-images.githubusercontent.com/40220712/128302050-212dd7ec-7efa-46bd-8c7e-ba3a00003bf7.png)

Thanks for the help!

**Reply by rayluo** [reply]
> I tried running this [repo](https://github.com/Azure-Samples/ms-identity-python-flask-webapp-authentication) as well and it returned the *same thing*, ......
> 
> Here's the snap of the error,
> ![image](https://user-images.githubusercontent.com/40220712/128302050-212dd7ec-7efa-46bd-8c7e-ba3a00003bf7.png)
> 

But this AADSTS9002325 is different than your earlier message's AADSTS9002327. They are not the same thing. You may create a new conversation on the new topic, but in this conversation, let's stick with the [first sample](https://github.com/Azure-Samples/ms-identity-python-webapp) and focus on the AADSTS9002327 error first. We will take a look soon. Meanwhile, have you tried to register your app as *NOT* a Single-Page Application? Python-powered apps are typically not SPAs.

**Reply by yancahila** [reply]
Alright! I just considered them similar since they're both pertaining to CORS but got your point. 

Ahm no I haven't tried that. That option doesn't seem to be available on Azure App Registration but I'll try to have a look on this. Tho, I have a workaround on this, I created my own http_client overriding the one created by the msal module and I just added the Origin header on the client and the app began working fine.

**Reply by rayluo** [reply]
@ian-cahila-maersk , your proposal/workaround of customizing Origin could be unnecessary in the first place. Have you tried my [previous suggestion 26 days ago](https://github.com/AzureAD/microsoft-authentication-library-for-python/discussions/392#discussioncomment-1133376), to either confirm or rule out that cause?

> Meanwhile, have you tried to register your app as NOT a Single-Page Application? Python-powered apps are typically not SPAs.

**Reply by rayluo** [reply]
In fact, I can reproduce your "AADSTS9002327: Tokens issued for the 'Single-Page Application' client-type may only be redeemed via cross-origin requests" error, ***if*** I deliberately configure my web app as a SPA app (and also skipping its CLIENT_SECRET in my app implementation) - but those were the wrong way to configure/use a web app.

If this was indeed your configuration, I'd suggest you to delete the SPA section, and configure your app as a Web app (or a Desktop app, for that matter). Or you could still have your SPA configuration and your Web/Desktop app configuration co-exist, as long as their redirect_uri are different than each other.

![SPA vs Desktop vs Web](https://user-images.githubusercontent.com/821550/131415905-00fbc494-3d60-43a0-a8c3-e59b7083b18c.png)

**Reply by yancahila** [reply]
Yes, it worked! My fault is that I was hesitant to make changes on the current implementation in the organization. I added the redirect_uri in the web app section and it still works fine and I agree with what you said about what I did as an incorrect way of doing this. I appreciate your help @rayluo. Thanks!

**Comment by rayluo** [comment, marked as answer]
(The following content was buried deep down [here in the lengthy conversation](https://github.com/AzureAD/microsoft-authentication-library-for-python/discussions/392#discussioncomment-1256665), and we copied it here for visibility.)

... I can reproduce your "AADSTS9002327: Tokens issued for the 'Single-Page Application' client-type may only be redeemed via cross-origin requests" error, ***if*** I deliberately configure my web app as a SPA app (and also skipping its CLIENT_SECRET in my app implementation) - but those were the wrong way to configure/use a web app.

If this was indeed your configuration, I'd suggest you to delete the SPA section, and configure your app as a Web app (or a Desktop app, for that matter). Or you could still have your SPA configuration and your Web/Desktop app configuration co-exist, as long as their redirect_uri are different than each other.

![SPA vs Desktop vs Web](https://user-images.githubusercontent.com/821550/131415905-00fbc494-3d60-43a0-a8c3-e59b7083b18c.png)
